gnd/src/main/java/com/google/android/gnd/persistence/local/LocalDataStore.java

```diff
@@ -28,8 +37,27 @@
 public interface LocalDataStore {
   /**
    * Applies the specified {@link FeatureMutation} to the local data store, appending the mutation
-   * to the local queue for remote sync. It is assumed to be a new, unsaved instance, and as such
-   * its {@code changeId} is ignored.
+   * to the local queue for remote sync.
    */
   Completable applyAndEnqueue(FeatureMutation mutation);
+
+  /**
+   * Applies the specified {@link RecordMutation} to the local data store, appending the mutation to
+   * the local queue for remote sync.
+   */
+  Completable applyAndEnqueue(RecordMutation mutation);
+
+  /**
+   * Returns a long-lived stream that emits the full set of features for a project on subscribe, and
+   * continues to return the full set each time a feature is added/changed/removed. The full set is
+   * returned rather than deltas for simplicity and to implement a fully reactive UI in which each
+   * update is idempotent.
```

Comment:
Sorry for the lengthy comment! 

I understand the first part of this comment--I can see how returning the full set is simpler than returning deltas--but I don't fully understand the second part of this comment. It seems like the UI layer is actually *less* reactive by operating on the entire set? To me, reactivity in this case would suggest that the UI reacts to changes to the underlying data structure, in this case the set, i.e. the UI is set up to "react" to alterations to the set, the flowable would simply propagate these changes to the UI. 

I'm wondering if this stream is at the right level of granularity. In general I'm suspicious of streams that emit iterable data structures for reasons other than semantic correspondence. For example, a stream of cards in a deck is different semantically from a stream of decks of cards where these decks differ, e.g. by language. That seems like a valid distinction and a valid place to use a stream that emits something that's iterable (decks) because we care about some property that emerges at the level of the deck (language). 

I think this case is actually one of the former--we are really dealing with cards (features) here, we care about differences at the card level and we are only using a single deck, not comparing decks (e.g. when do we add a new card into a single deck, when do we remove a card from the deck, when do we change the face of a card in the deck), not at the deck level (when do we get a new deck, when do we change the language of an entire deck, when do we remove an entire deck--in our terms this would be something akin to caring about when *all* features are simultaneously removed from a project, when all features are simultaneously updated etc.).

If I understand this correctly, as this is set up we will either wind up re-rendering a lot of features already rendered, or have to execute a set operation (complement) to render only those features that changed. Instead of reacting to changes on the set--I think cases like this under-utilize RX...and I think we're mostly only nudged in this direction by the design of Live Data (at least on the UI side). One of the major benefits of RX is turning structures we'd normally explicitly iterate over into declaratively established dependencies between streams and functions. We lose this benefit using the approach here, since we wind up having to iterate over the stream's contents if we don't want to do unnecessary work. (obviously, this unnecessary work is negligible for projects with few features, which makes this a non-issue for the majority of cases. But what happens if I have a project (assuming this is possible) with 10,000 features? Suddenly it seems like only rendering those features that actually changed would be much more efficient, which pushes us into iterating over the set, when we could just let RX handle this and react to feature-level emissions).

All that said, idempotency is typically a good thing, and I'm not privy to the other considerations that went into this approach that might make my musings above foolish.

Reply:
Thanks for the comment! I while back I had a great conversation with @yigit about this which changed our thinking and resulted in the current implementation. 

My understanding was that to ensure LiveData works even when the UI misses some emissions, and that all subscribing UIs are consistent, LiveData imposes that UIs are state-driven. In other words, emissions in a stream represent the latest state (e.g., here's the current list of things to show), not an event to be applied to existing state (add/remove/update such and such a thing in your list).

To properly implement the event/action/delta based model, you'd need to maintain state at some point in order to have something for events to mutation. Given Android Lifecycles, it's very hard to guarantee integrity and consistency of cached states. Also, assuming the stream always emits the latest state (e.g. list of pins being show) makes it a lot easier to reason about the current state of the UI.

Although intuitively sending the full set of UI elements on each change seems inefficient, in practice the diffing done at view level turns out the work quite well. In rare cases where it's too costly, workarounds can be found.

@yigit, please cmiiw!